hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+    -------
+    vres : list
+        The transfer resistance at each segment of the section.
+    """
+    electrode_pos = np.array(electrode_pos)  # electrode position to Numpy
+
+    sec_start = np.array([section.x3d(0), section.y3d(0), section.z3d(0)])
+    sec_end = np.array([section.x3d(1), section.y3d(1), section.z3d(1)])
+    sec_vec = sec_end - sec_start
+
+    # NB segment lengths aren't equal! First/last segment center point is
+    # closer to respective end point than to next/previous segment!
+    # for nseg == 5, the segment centers are: [0.1, 0.3, 0.5, 0.7, 0.9]
+    seg_ctr = np.zeros((section.nseg, 3))
+    line_lens = np.zeros((section.nseg + 2))
+    for ii, seg in enumerate(section):
+        seg_ctr[ii, :] = sec_start + seg.x * sec_vec
+        line_lens[ii + 1] = seg.x * section.L
+    line_lens[-1] = section.L
+    line_lens = np.diff(line_lens)
+    first_len = line_lens[0]
+    line_lens = np.array([first_len] + list(line_lens[2:]))
+
+    if method == 'psa':
+
+        # distance from segment midpoints to electrode
+        dis = norm(np.tile(electrode_pos, (section.nseg, 1)) - seg_ctr,
+                   axis=1)
+
+        # To avoid very large values when electrode is placed close to a
+        # segment junction, enforce minimal radial distance
+        dis = np.maximum(dis, min_distance)
+
+        phi = 1. / dis
+
+    elif method == 'lsa':
+        # From: Appendix C (pp. 137) in Holt, G. R. A critical reexamination of
+        # some assumptions and implications of cable theory in neurobiology.
+        # CalTech, PhD Thesis (1998).
+        #
+        #                      Electrode position
+        #   |------ L --------*
+        #                 b / | R
+        #                 /   |
+        #   0==== a ====1- H -+
+        #
+        # a: vector oriented along the section
+        # b: position vector of electrode with respect to section end (1)
+        # H: parallel distance from section end to electrode
+        # R: radial distance from section end to electrode
+        # L: parallel distance from section start to electrode
+        # Note that there are three distinct regimes to this approximation,
+        # depending on the electrode position along the section axis.
+        phi = np.zeros(len(seg_ctr))
+        for idx, (ctr, line_len) in enumerate(zip(seg_ctr, line_lens)):
+            start = ctr - line_len * sec_vec / norm(sec_vec)
+            end = ctr + line_len * sec_vec / norm(sec_vec)
+            a = end - start
+            norm_a = norm(a)
+            b = electrode_pos - end
+            # projection: H = a.cos(theta) = a.dot(b) / |a|
+            H = np.dot(b, a) / norm_a  # NB can be negative
+            L = H + norm_a
+            R2 = np.dot(b, b) - H ** 2  # NB squares
+
+            # To avoid very large values when electrode is placed (anywhere) on
+            # the section axis, enforce minimal perpendicular distance
+            R2 = np.maximum(R2, min_distance ** 2)
```

Comment:
Maybe just think about what `min_distance` is doing before merging @jasmainak and @ntolley 

I think this'll come up again when repositioning the cells, so I wouldn't call it a show-stopper for merging this PR. Just something to keep in mind.

Reply:
Well the math gets "screwed up" either way, but this is to avoid division by zero. The situation is clearer in PSA, where the euclidean distance is restricted to be `> min_distance`. I noticed that LFPy and NetPyNE use a limit on R, too. We can return to this: it will resurface when talking about physical distances. Right now it "works".

Reply:
If the electrode is so close to section axis, wouldn't the voltage simply be the membrane potential? I guess accessible with `seg.v` ?